### DISCUSSION TITLE
Farfield of Dielectric Disc

### DISCUSSION THREAD

**Discussion by rodechamp** [NanoComp/meep#2056] [Q&A] [answered]
Hello meep community!

I'm trying to calculate the farfield of a whispering-gallery mode (WGM) in a disc. However,  the resulting farfield is nowhere near the expected circular symmetry. Did I do something wrong or is this an artifact of the numerical calculation?

For that I use meeps harminv to obtain the resonant frequency of the mode. Afterwoards I use the obtained frequency and a very low frequency width of the gaussian source to excite the mode. The following image shows the intensity of $E_z$ on a logarithmic scale, with the PMLs surrounding the cell omitted. It can be seen that the mode has been excited succesfully, with the lines of minima in the center being a remnant of the source position. The darker area notes the position of the disc with a refractive index of $n = 3$.

![pattern_start](https://user-images.githubusercontent.com/104635807/166661246-349328e8-8ce3-476a-9de3-2c4ff9ab3fe0.png)

After the source is turned off, the nearfield box is added in order to perform the near-to-farfield transformation. After the simulation is run another 100 periods, the farfield is calculated. The results can be seen below.

![farfield](https://user-images.githubusercontent.com/104635807/166661258-afcc7ced-3516-417d-a024-a54674b1c8f7.png)

The farfield is not even close to showing the expected symmetry. The maxima at 0 and 180 degrees are due to the source position. I tried several things (longer simulation time, varying the resolution, frequencies). And I am out of ideas. It puzzles me how the horizontal mode in the cavity can dominate the farfield, when its field intensity is three orders of magnitude below the WGM. Even if its $Q$-factor is far lower than that of the WGM, I would expect the find more of the WGM in the farfield. In addition,  in the picture of the field you see a four-fold symmetry, which doesn't make any sense for a circular microcavity.

I would appreciate any help, as I am out of ideas! For completeness sake, the code for calculating the fields in the plots is found below. Thanks in advance!

```
'''
File with code for meep forum question about the farfield
of a dielectric disc.
'''

import matplotlib
import matplotlib.pyplot as plt
import meep as mp
import numpy as np


#########################################
# 1. set general parameters 
#########################################


RESOLUTION = 100.
DPML = 1.

# geometry params
RADIUS = 1.
PAD = 0.05      # distance from nearfield box to PML
DIST_NFB = 1.5  # distance from radius to the nearfield box
INDEX = 3.

# frequency information
FCEN = 1.00
DF = 0.01

# determine cell size

dx = ( RADIUS + DIST_NFB + PAD ) * 2.
dy = dx

cell_size_nodpml = mp.Vector3(dx,dy)
cell_size = mp.Vector3(dx + 2.*DPML, dy + 2.*DPML)



#########################################
# 2. create geometry
#########################################


geometry = []
disc = mp.Cylinder(
    radius=RADIUS,
    material=mp.Medium(index=INDEX),
    height=1.
    )
geometry.append(disc)


#########################################
# 3. create source
#########################################

sources = []

source_xpos = 0.95 * RADIUS
source_center = mp.Vector3(source_xpos, 0)

src = mp.Source(
    mp.GaussianSource(
        FCEN,
        fwidth=DF,
        ),
    component=mp.Ez,
    center=source_center,
    )
sources.append(src)

#########################################
# 4. simulation
#########################################

print('\nrunning sources...\n')

# def simulation
sim = mp.Simulation(
    cell_size=cell_size,
    resolution=RESOLUTION,
    geometry=geometry,
    sources=sources,
    boundary_layers=[mp.PML(DPML)]
    )

# run simulation
sim.run(
    until_after_sources=100/FCEN # run for 100 periods
    )

#########################################
# 5. saving output (until after sources)
#########################################

dielectric = sim.get_array(
    component=mp.Dielectric,
    size=cell_size_nodpml,
    center=mp.Vector3()
    )
dielectric = np.transpose(dielectric)

Ez = sim.get_array(
    component=mp.Ez,
    size=cell_size_nodpml,
    center=mp.Vector3()
    )
Ez = np.transpose(Ez)

# plotting

fig, ax = plt.subplots()

ax.imshow(
    dielectric,
    cmap='binary',
    origin='lower'
    )

im = ax.imshow(
    np.abs(Ez),
    origin='lower',
    alpha=0.9,
    norm=matplotlib.colors.LogNorm()
    )

plt.colorbar(im,label=r'$|E_z|$')

ax.set_axis_off()

fig.savefig('pattern_start.png',dpi=100,bbox_inches='tight')


#########################################
# 6. define nearfield box
#########################################

print('\nrunning farfields...\n')


nfb_dx = (RADIUS + DIST_NFB) * 2. # cell in nearfield-box
nfb_dy = nfb_dx

nearfield_box = sim.add_near2far(
    FCEN, 0, 1,
    mp.Near2FarRegion( # up
        center=mp.Vector3(0,+0.5*nfb_dy), size=mp.Vector3(nfb_dx,0), weight=+1
        ),
    mp.Near2FarRegion( # down
        center=mp.Vector3(0,-0.5*nfb_dy), size=mp.Vector3(nfb_dx,0), weight=-1
        ),
    mp.Near2FarRegion( # right
        center=mp.Vector3(+0.5*nfb_dx,0), size=mp.Vector3(0,nfb_dy), weight=+1
        ),
    mp.Near2FarRegion( # left
        center=mp.Vector3(-0.5*nfb_dx,0), size=mp.Vector3(0,nfb_dy), weight=-1
        )
    )

flux_box = sim.add_flux(FCEN, 0, 1,
    mp.FluxRegion( # up
        center=mp.Vector3(0,+0.5*nfb_dy), size=mp.Vector3(nfb_dx,0), weight=+1
        ),
    mp.FluxRegion( # down
        center=mp.Vector3(0,-0.5*nfb_dy), size=mp.Vector3(nfb_dx,0), weight=-1
        ),
    mp.FluxRegion( # right
        center=mp.Vector3(+0.5*nfb_dx,0), size=mp.Vector3(0,nfb_dy), weight=+1
        ),
    mp.FluxRegion( # left
        center=mp.Vector3(-0.5*nfb_dx,0), size=mp.Vector3(0,nfb_dy), weight=-1
        )
    )

#########################################
# 7. run
#########################################

sim.run(
    mp.Harminv(mp.Ez, source_center, FCEN, DF), # return frequencies to prove that FCEN is resonant
    until=500/FCEN # run for another 100 periods
    )


#########################################
# 8. calculate far field
#########################################

near_flux = mp.get_fluxes(flux_box)[0]

r = 1000/FCEN      # half side length of far-field square box OR radius of far-field circle
res_ff = 1         # resolution of far fields (points/μm)
far_flux_box = (
    + nearfield_box.flux(mp.Y, mp.Volume(center=mp.Vector3(y=r), size=mp.Vector3(2*r)), res_ff)[0]
    - nearfield_box.flux(mp.Y, mp.Volume(center=mp.Vector3(y=-r), size=mp.Vector3(2*r)), res_ff)[0]
    + nearfield_box.flux(mp.X, mp.Volume(center=mp.Vector3(r), size=mp.Vector3(y=2*r)), res_ff)[0]
    - nearfield_box.flux(mp.X, mp.Volume(center=mp.Vector3(-r), size=mp.Vector3(y=2*r)), res_ff)[0]
    )

npts = 1000         # number of points in [0,2*pi) range of angles
angles = 2*np.pi/npts*np.arange(npts)

field_vectors = []
for angle in angles:
    ff = sim.get_farfield(
        nearfield_box,
        mp.Vector3(r*np.cos(angle), r*np.sin(angle)))
    field_vectors.append(ff)
    
field_vectors = np.transpose(field_vectors) # make the columns to rows
Ex, Ey, Ez, Hx, Hy, Hz = field_vectors

# P = E* x H
Ex, Ey, Ez = np.conj([Ex, Ey, Ez])
Px = np.real(Ey*Hz-Ez*Hy)
Py = np.real(Ez*Hx-Ex*Hz)
Pr = np.sqrt(np.square(Px)+np.square(Py))

far_flux_circle = np.sum(Pr)*2*np.pi*r/len(Pr)

print("flux:, {:.6f}, {:.6f}, {:.6f}".format(near_flux,far_flux_box,far_flux_circle))

#########################################
# 9. plotting farfield
#########################################

fig, ax = plt.subplots(
    subplot_kw={'projection' : 'polar'}
    )

ax.plot(angles, Pr/np.max(Pr))


fig.savefig('farfield.png',dpi=100,bbox_inches='tight')

#########################################
# 10. saving output (after farfield calculation)
#########################################

Ez = sim.get_array(
    component=mp.Ez,
    size=cell_size_nodpml,
    center=mp.Vector3()
    )
Ez = np.transpose(Ez)

fig, ax = plt.subplots()

ax.imshow(
    dielectric,
    cmap='binary',
    origin='lower'
    )

im = ax.imshow(
    np.abs(Ez),
    origin='lower',
    alpha=0.9,
    norm=matplotlib.colors.LogNorm()
    )

plt.colorbar(im,label=r'$|E_z|$')

ax.set_axis_off()

fig.savefig('pattern_end.png',dpi=100,bbox_inches='tight')
plt.close(fig)
```

**Comment by smartalecH** [comment]
(cc @hammy4815)

**Comment by ianmatthewhammond** [comment]
Looking through the code, I would suggest trying a few things at first:

1. Try running for longer (just run it for a while longer and see if fields converge to same shape)
2. Try various forms for Pr. (E x H*) for example, and try calculating Pr as (Px * x + Py * y) / sqrt(x**2+y**2) or Px * y + Py * x … depending on your vector orientation if you create a mesh grid to do this. I believe this is the formulation you want for a vector field on the grid. 
3. See what happens if you include your farfield in the source decay region of time for more vs less time. May give an idea of which direction to take it to balance out the symmetry


Let me know if none of these help and I can try and come up with some more suggestions.

**Reply by rodechamp** [reply]
Thanks for the ideas! I already tried some of your suggestions, but nothing come of it.

1. Running the sim longer yields a slightly more isotropic far-field, however the maxima at about 90 and 270 degrees increase in intensity, but become more and more noisy.
2. I will try this. But I don't quite understand how the step is viable, as the flux through the near-field surface is only preserved if Pr is calculated as above?
3. I tried that, but then the source appears to dominate the far-field, as the cavity only radiates very little energy due to the high Q factor. See my answer to oskooi.

**Reply by ianmatthewhammond** [reply]
Pr is defined as the component of P pointing radially at any angle theta. You can imagine a vector field P that only points in the x direction P=<5,0> for example. According to your definition for Pr, sqrt(Px^2+Py^2) gives 5 at every point. But this is not correct. The radial component of P, (Pr), should be highest at angles radially pointing in x such as 0 and 180 on the unit circle and lowest pointing in y such as 90 and 270. This example vector field thus should be zero at 90 and 270 because the field only points in x.

To calculate the radial part, you want Pr=Px*cos(theta)+Py*sin(theta). If you plug in the example vector field from above into this equation, you will see it matches what we expect the radial component to be. cos(theta) can be written as x/sqrt(x^2+y^2) which is where my last response formulation came from, but either should work.

I'm not sure if this will solve your problem, but i think it's worth trying. I would've expected a symmetry in your formulation as well, but maybe because you do not include Pz that might account for it...although at the level plane I wouldn't expect much or any Pz. You should also try looking at |E| because that should also have a symmetry in the lobes. Like other comments have mentioned, I think it's more likely to do with how long and when the source is included in the farfield calculation.

**Reply by rodechamp** [reply]
Ah thanks, that makes sense. I tried it and as you expected, it yielded pretty much the same results as taking |Pr|.

Right now I'm pretty sure the problem is the high $Q$-factor, as the farfield calculation works properly for modes which decay faster.

**Comment by oskooi** [comment]
Actually, I think the problem here is that you could just be measuring noise in the near fields. This is because you first run for 100 periods after the sources turns off *and then* add your near-field monitors and run for another 500 periods. Depending on the quality factor (Q) of this particular cavity, the fields may have decayed away after the first run to such an extent that the near fields measured as part of the second may be practically zero.

The fix is straightforward: since you have already precomputed the resonant frequency, run just once and add the near-field monitors *before* you start the run.

**Reply by rodechamp** [reply]
Hello, thanks for your response! Adding the NF monitors before running was the first thing I tried, as was done in the [tutorial](https://meep.readthedocs.io/en/latest/Python_Tutorials/Near_to_Far_Field_Spectra/). But in that case the source itself dominates the far-field... Modifying the code above like you said gives:

![farfield](https://user-images.githubusercontent.com/104635807/166915929-179a7ba8-b481-41d2-94a3-ae287168bc2e.png)

The Q factor of the mode is very high so the field inside the cavity is not dissipated at the end of the simulation. The high Q factor also makes it near impossible to simulate until the fields are sufficiently decayed (as in the tutorial), as it can take more than a week to calculate the far-field this way.

**Comment by rodechamp** [comment, marked as answer]
Thanks again for all the ideas! After playing around a bit more I found that the farfield calculation works properly if the modes have lower $Q$-factors. Probably the near- to farfield calculation does something due to a lot of noise at the nearfield box, as little to no field leaks out of the cavity.

For this, compare the mode with $f=1.172$ and $Q\approx10^6$

<img src="https://user-images.githubusercontent.com/104635807/172829147-1035302c-0975-4671-b835-dc3ac6162be1.png" alt="pattern_highQ" width="500"/>
<img src="https://user-images.githubusercontent.com/104635807/172829195-5f68d699-8874-48bb-99e9-a3cd92202c70.png" alt="farfield_highQ" width="500"/>


with $f = 0.779$ and $Q\approx10^3$

<img src="https://user-images.githubusercontent.com/104635807/172829315-71af668d-f1a4-421e-9c82-39107175ed1c.png" alt="pattern_highQ" width="500"/>
<img src="https://user-images.githubusercontent.com/104635807/172829331-b51dcb25-04a7-4fff-b058-964d1cb6a2fd.png" alt="farfield_highQ" width="500"/>

The data above is calculated as in the code in my original question, where the farfield is recorded only after the source is turned off. After the sources, the simulation is run for 1000 timesteps. The latter shows exactly what analytical results suggest. The frequencies and $Q$-factors are calculated in advance using the `harminv` feature.

**Reply by truonghoailinh47** [reply]
> Thanks again for all the ideas! After playing around a bit more I found that the farfield calculation works properly if the modes have lower Q-factors. Probably the near- to farfield calculation does something due to a lot of noise at the nearfield box, as little to no field leaks out of the cavity.
> 
> For this, compare the mode with f=1.172 and Q≈106
> 
> <img alt="pattern_highQ" width="500" src="https://user-images.githubusercontent.com/104635807/172829147-1035302c-0975-4671-b835-dc3ac6162be1.png">
> 
> <img alt="farfield_highQ" width="500" src="https://user-images.githubusercontent.com/104635807/172829195-5f68d699-8874-48bb-99e9-a3cd92202c70.png">
> 
> with f=0.779 and Q≈103
> 
> <img alt="pattern_highQ" width="500" src="https://user-images.githubusercontent.com/104635807/172829315-71af668d-f1a4-421e-9c82-39107175ed1c.png">
> 
> <img alt="farfield_highQ" width="500" src="https://user-images.githubusercontent.com/104635807/172829331-b51dcb25-04a7-4fff-b058-964d1cb6a2fd.png">
> 
> The data above is calculated as in the code in my original question, where the farfield is recorded only after the source is turned off. After the sources, the simulation is run for 1000 timesteps. The latter shows exactly what analytical results suggest. The frequencies and Q-factors are calculated in advance using the `harminv` feature.

Hello. Can you guide me how to adjust f and Q factor in your code to get the result that is same f=1.172 and Q≈106 of yours? I have just learned MEEP python, I have a little bit missunderstand. Thank a lot.
